File: leetcode/golang/213.打家劫舍II.go
func rob(nums []int) int {
    n := len(nums);
    if n == 1 {
        return nums[0];
    }
    left := rob2(nums[:n - 1]);
    right := rob2(nums[1:]);
    if left > right {
        return left;
    }
    return right;
}

func rob2(nums []int) int {
    i_1, i_2 := 0, 0;
    for _, num := range nums {
        tmp := i_1;
        if num + i_2 > i_1 {
            i_1 = num + i_2;
        }
        i_2 = tmp;
    }
    return i_1;
}

File: leetcode/python/213.打家劫舍II.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1: return nums[0]
        return max(self.rob2(nums[: n - 1]), self.rob2(nums[1:]))

    def rob2(self, nums: List[int]) -> int:

        i_1, i_2 = 0, 0
        for num in nums:
            i_2, i_1 = i_1, max(num + i_2, i_1)
        return i_1